migration,server: plumb in initial version to the migration manager

irfansharif opened the pull request:
It'll let us return early, and makes the manager "more functional" in its
behavior. 

We also promote the clusterversion type to a first-class citizen in the
external facing API for pkg/migration. This package concerns itself with
migrations between cluster versions and we should have our API reflect as much.

The proto changes are safe, we haven't had a major release with the previous
proto definitions.

Release note: None

irfansharif committed:
migration,server: plumb in initial version to the migration manager

It'll let us return early, and makes the manager "more functional" in
its behavior. We should also be using the ClusterVersion type when
talking about migrations, so we make that change here.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57118)
<!-- Reviewable:end -->

irfansharif committed:
migration: promote clusterversion to first-class citizen...

in the external facing API for pkg/migration.

The migration package concerns itself with migrations between cluster
versions, so we should have our API reflect as much. The proto changes
are safe, we haven't had a major release with the previous proto
definitions.

Release note: None

irfansharif committed:
clusterversion: rename proto pkg name for cleaner import path

Release note: None

irfansharif committed:
clusterversion: document process for introducing cluster versions

And for deleting old ones. This stuff is endlessly confusing.

Release note: None

irfansharif committed:
clusterversion: move version comments next to definition

Jump to symbol for a given version tag will now group together the
intent of the version itself (instead of expecting readers to know to
grep for the same version in the block below).

Release note: None

tbg reviewed (approved):
> Reviewed 5 of 5 files at r1, 2 of 2 files at r2, 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57118)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/migration/manager.go, line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57118#-MN7wGHV7rLAOutSEBft:-MN7wGHW1MZcNbgmoo8l:bnu8ynl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5ee31b87f5a1635360b7c699910d3237f63360e/pkg/migration/manager.go#L254)):*
> ```Go
> 	for _, cv := range cvs {
> 		h := &Helper{Manager: m}
> 		clusterVersion := cv
> ```

What good is this? Let's avoid unnecessary aliasing. Why not just `for _, clusterVersion := range cvs`?


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
Thanks!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57118#-:-MNAqR1G8383uY6-vlKF:b-voiyyw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/migration/manager.go, line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57118#-MN7wGHV7rLAOutSEBft:-MNAqKzw63ABKSfhyuIi:b-wewgvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5ee31b87f5a1635360b7c699910d3237f63360e/pkg/migration/manager.go#L254)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

What good is this? Let's avoid unnecessary aliasing. Why not just `for _, clusterVersion := range cvs`?
</blockquote></details>

Done (was just a brain fart).


<!-- Sent from Reviewable.io -->
